cmd/flags.go

```diff
@@ -102,8 +103,11 @@ func CreateFlags(defaultPath string) []cli.Flag {
 			Usage: "By setting this flag to true, disables the need to wait the propagation of the TXT record to all authoritative name servers.",
 		},
 		&cli.StringSliceFlag{
-			Name:  "dns.resolvers",
-			Usage: "Set the resolvers to use for performing recursive DNS queries. Supported: host:port. The default is to use the system resolvers, or Google's DNS resolvers if the system's cannot be determined.",
+			Name: "dns.resolvers",
+			Usage: "Set the resolvers to use for performing (recursive) CNAME resolving and apex domain determination." +
+				" For DNS challenge verification, the authoritative DNS server is queried directly." +
+				" Supported: host:port." +
+				" The default is to use the system resolvers, or Google's DNS resolvers if the system's cannot be determined.",
```

Comment:
I'm still not clear what this means. What does setting this value gain a user?

Reply:
(I'll need to provide a bit of context, please bear with me for a moment. I'll try to be brief, but that also requires to simplify things. This is the ELI5 version.)

---

The ACME protocol involves a few steps of verification, in order to issue you a TLS certificate. One of the earliest steps is to ask the ACME provider (usually Let's Encrypt), if `lego` can have such a certificate for `www.example.com`, to which Let's Encrypt answers, "sure, if you can prove you own that domain; here is a token and place it someplace only accessible to owners".

Proving ownership can happen in different ways, but the bulk of the code in this repository is concerned with the "DNS challenge" (dns-01). This means, `lego` must place the aforementioned token as TXT record on `_acme-challenge.www.example.com`.

Now, DNS being the redundant and resilient system it is, is not known for fast propagation of changes. Layers of caching servers are setup between you and the name server responsible to deliver you that TXT record (on your PC, your router, your ISP have certainly installed a recursive resolver). Each layer may (or may not!) obey the caching time for certain DNS answers, even for negative answers ("I don't know about `_acme-challenge.www.example.com`, and the authoritative name server has told me to save this answer for 600 seconds").

This makes the verifying step tedious. `lego` could skip this step, and leave it to Let's Encrypt. But should setting up the TXT record fail (or just take a moment), Let's Encrypt may just respond with "you didn't setup the token, you don't get a certificate", and deduct one point from your request quota (remember that Let's Encrypt rate limits your account to 300 new orders per 3 hours). If you try and retry and retry to obtain a certificate, your quota may be exhausted quicker than you think.

The next less naïve approach would be to wait a moment. But that begs the question: for how long? Should every user generally suffer from overly long waiting periods, just because in some cases the propagation takes a bit longer? Certainly not.

So... what does `lego` do? Easy: We're querying the authoritative DNS server directly, avoiding any caches. The not-so-easy part is determining which server that is.

For a domain like `example.com`, we can just lookup the SOA record for `example.com`, and extract the MNAME from it (in DNS lingo, `example.com` is the apex domain for `example.com`, because it contains the SOA record). This however gets a little more complicated for domains like `zeus.fra.ger.eu.cdn.example.com`, where the whole `*.cdn.example.com` zone might be delegated to another server (the apex domain might be `cdn.example.com`, not `example.com`).

Here, `lego` can only ask first `zeus.fra.ger.eu.cdn.example.com` for its SOA record, then `fra.ger.eu.cdn.example.com`, etc., until it finds the authoritative name server.

The final step is to query that name server, "hey what's the TXT record for `_acme-challenge.zeus.fra.ger.eu.cdn.example.com`? If it doesn't provide the correct answer, `lego` will wait a moment (`--dns-timeout`) and try again (I forgot how often).

(The story then continues with a similar verification on Let's Encrypt's side, and the exchange of the TLS certificate, should LE be satisfied.)

---

What does setting [the `--dns.resolvers`] value gain a user?

Since `lego`, by default, uses the system's DNS resolver to determine the apex domain, this may lead to a wrong answer. Within data centers you might have DNS setup in a way which gives you a different answer than queries in from the outside world ("split DNS", "hidden master").

A false-positive here is not better than the first approach described above (i.e. let Let's Encrypt figure it out). A false-negative here could mean that your system resolver is not able at all to get the `_acme-challenge` record - and `lego` would wait and wait and finally fail.

Another reason can be that you don't trust your system's/router's/ISP's name server. I know from some residential ISPs here in Germany where in the past they wanted to be helpful and present users with a "Did you mean alice-dsl.de? Here are some ads"-page, if the end-user had a typo when entering a domain into the browsers URL bar (/rant). Of course such a behaviour breaks `lego`'s verification process as well.

In such situations, you need to instruct `lego` to use a specific (set of) known-good resolver(s) (ideally but not necessarily on the public internet). The `--dns.resolvers` flag is for this specific requirement.